OpenRA.Mods.Common/Traits/DamagedByTerrain.cs

```diff
@@ -32,68 +32,35 @@ class DamagedByTerrainInfo : ConditionalTraitInfo, Requires<IHealthInfo>, Requir
 		[Desc("Terrain types where the actor will take damage.")]
 		public readonly string[] Terrain = { };
 
-		[Desc("Percentage health below which the actor will not receive further damage.")]
-		public readonly int DamageThreshold = 0;
-
-		[Desc("Inflict damage down to the DamageThreshold when the actor gets created on damaging terrain.")]
-		public readonly bool StartOnThreshold = false;
-
-		public override object Create(ActorInitializer init) { return new DamagedByTerrain(init.Self, this); }
+		public override object Create(ActorInitializer init) { return new DamagedByTerrain(this); }
 	}
 
-	class DamagedByTerrain : ConditionalTrait<DamagedByTerrainInfo>, ITick, ISync, INotifyAddedToWorld
+	public class DamagedByTerrain : ConditionalTrait<DamagedByTerrainInfo>, ITick, ISync
 	{
-		readonly IHealth health;
-
-		[Sync]
+		bool onDamagingTerrain;
+		CPos cachedLocation;
 		int damageTicks;
 
-		[Sync]
-		int damageThreshold;
-
-		public DamagedByTerrain(Actor self, DamagedByTerrainInfo info)
-			: base(info)
-		{
-			health = self.Trait<IHealth>();
-		}
-
-		void INotifyAddedToWorld.AddedToWorld(Actor self)
-		{
-			if (!Info.StartOnThreshold || IsTraitDisabled)
-				return;
-
-			var safeTiles = 0;
-			var totalTiles = 0;
-			foreach (var kv in self.OccupiesSpace.OccupiedCells())
-			{
-				totalTiles++;
-				if (!Info.Terrain.Contains(self.World.Map.GetTerrainInfo(kv.Cell).Type))
-					safeTiles++;
-			}
-
-			if (totalTiles == 0)
-				return;
-
-			// Cast to long to avoid overflow when multiplying by the health
-			damageThreshold = (int)((Info.DamageThreshold * (long)health.MaxHP + (100 - Info.DamageThreshold) * safeTiles * (long)health.MaxHP / totalTiles) / 100);
-
-			// Actors start with maximum damage applied
-			var delta = health.HP - damageThreshold;
-			if (delta > 0)
-				self.InflictDamage(self.World.WorldActor, new Damage(delta, Info.DamageTypes));
-		}
+		public DamagedByTerrain(DamagedByTerrainInfo info)
+			: base(info) { }
 
 		void ITick.Tick(Actor self)
 		{
-			if (IsTraitDisabled || health.HP <= damageThreshold || --damageTicks > 0)
+			if (IsTraitDisabled || --damageTicks > 0)
 				return;
 
 			// Prevents harming cargo.
 			if (!self.IsInWorld)
 				return;
 
-			var t = self.World.Map.GetTerrainInfo(self.Location);
-			if (!Info.Terrain.Contains(t.Type))
+			if (cachedLocation != self.Location)
```

Comment:
This will regress the case where tiberium grows underneath where infantry are standing. For the sake of getting this PR merged without further delay IMO we should just remove the caching.

Reply:
Ah right, that is why I wanted to look into `CellLayer.CellEntryChanged`. I think the issue here was that there are several layers (can I cover all?) and just using the one for resources might be too specific. (Even though it would work for our cases?)

Reply:
All the layers should end up modifying the same `Map.CustomTerrain`. We can have the `ActorMap` add a listener to that and then notify any actors on the changed cells. Straightforward in theory, but I suspect this will end up being big enough that we won't want to tackle it right now.

Reply:
Sounds good for a follow-up. Removed for now.